rust/src/detect/xbits.rs

```diff
@@ -0,0 +1,224 @@
+    }
+    match tracker[0].trim() {
+        "track" => {}
+        _ => {
+            SCLogError!("xbits track keyword parsing failed");
+            return Err(());
+        }
+    }
+    let tracker: u8 = match get_xbit_type(tracker[1].trim()) {
+        Ok(val) => val,
+        Err(e) => {
+            SCLogError!("xbits tracker parsing failed: {:?}", e);
+            return Err(());
+        }
+    };
+    let vartype = if tracker == DETECT_XBITS_TRACK_IPPAIR {
+        11 // VAR_TYPE_IPPAIR_BIT
+    } else {
+        8 // VAR_TYPE_HOST_BIT
+    };
+    let cmd: u8 = match get_xbit_cmd(args[0].trim()) {
+        Ok(val) => val,
+        Err(_) => {
+            SCLogError!("xbits cmd parsing failed");
+            return Err(());
+        }
+    };
+    let name = CString::new(args[1].trim()).unwrap();
+    if args.len() == 3 {
+        return Ok((cmd, name, tracker, DETECT_XBITS_EXPIRE_DEFAULT, vartype));
+    }
+    let expire: Vec<&str> = args[3].trim().split(' ').collect();
+    if expire.len() != 2 {
+        return Err(());
+    }
+    match expire[0].trim() {
+        "expire" => {}
+        _ => {
+            SCLogError!("xbits expire keyword parsing failed");
+            return Err(());
+        }
+    }
+    let expire: u32 = match expire[1].trim().parse() {
+        Ok(val) => val,
+        Err(_) => {
+            SCLogError!("xbits expire parsing failed");
+            return Err(());
+        }
+    };
+    if expire == 0 {
+        return Err(());
+    }
+    Ok((cmd, name, tracker, expire, vartype))
+}
+
+pub fn parse_xbits(arg: &str) -> Result<RSXBitsData, ()> {
+    let split_args: Vec<&str> = arg.split(',').collect();
+    let res;
+
+    match split_args.len() {
+        1 => match split_args[0] {
+            "noalert" => {
+                return Ok(RSXBitsData {
+                    cmd: get_xbit_cmd("noalert")?,
+                    name: std::ptr::null_mut(),
+                    tracker: 0,
+                    expire: 0,
+                    vartype: 0,
+                })
+            }
+            _ => {
+                SCLogError!("xbits noalert parsing failed");
+                return Err(());
+            }
+        },
+        3 | 4 => {
+            res = match evaluate_args(split_args) {
+                Ok(val) => val,
+                Err(_) => return Err(()),
+            };
+        }
+        e => {
+            SCLogError!("Erroneous number of arguments: {}", e);
+            return Err(());
+        }
+    }
+    Ok(RSXBitsData {
+        cmd: res.0,
+        name: (res.1).into_raw(),
+        tracker: res.2,
+        expire: res.3,
+        vartype: res.4,
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_xbits_parse(carg: *const c_char) -> *mut c_void {
+    if carg.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let arg = match CStr::from_ptr(carg).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
```

Comment:
In fact, this can be tidied up a bit if there is no special error handling to
```
if let Ok(arg) = CStr::from_ptr(carg).to_str() {
        if let Ok(detect) = parse_xbits(arg) {
            return Box::into_raw(Box::new(detect));
        }
    }
    std::ptr::null_mut();
```